### ISSUE TITLE
Generic interface

### ISSUE THREAD

**Issue by 00shiv** [stainless-steel/lapack#6]
Would it be a good idea to have a trait Lapack that is implemented for f32, f64, c32, c64 to allow generics and unify the 4 naming schemes?

**Comment by IvanUkhov** [comment]
Thanks for the suggestion! To be honest, I’m not sure if this is a good place for an additional level of abstraction. The [`blas`](https://crates.io/crates/blas) and [`lapack`](https://crates.io/crates/lapack) packages are meant to be skinny wrappers around what BLAS and LAPACK, respectively, offer.

On the other hand, it would be nice to get rid of the `s`, `d`, `c`, and `z` prefixes and simplify the usage of the packages. My primary concerns are: (a) the ability to avoid paying any runtime cost (zero-cost abstractions™) and (b) the ability to regenerate the contents of the packages by means of a script such as [this](https://github.com/stainless-steel/blas/blob/master/generate.py) one or [that](https://github.com/stainless-steel/lapack/blob/master/generate.py) one. The second aspect has been proven to be especially advantageous as the number of functions is huge, and, hence, it is highly error-prone to make any changes manually.

If you have any concrete proposal, it would be nice to have a look at patches of code for a function or two to see how it all plays out and fits into the scope of `blas` and `lapack`.

Thanks again!


Regards,
Ivan

**Comment by astraw** [comment]
I just published [nalgebra-lapack](https://github.com/astraw/nalgebra-lapack) which sits on top of nalgebra and lapack and defines traits HasSVD() and HasEigenvalue() and provides implementations for nalgebra matrices of supported type. I don't think it would be possible to write a generic implementation that would cover complex and real types because the LAPACK interface is different for those cases (e.g. the number of arguments to dgesvd and zgesvd is different). So instead I wrote macros to provide the implementation for each supported type. Perhaps when rust supports specialization for generics this could be revisited. In the meantime, perhaps this code, or the ideas behind it are useful to you.
